bundles/io/org.eclipse.smarthome.io.net.security.test/src/test/java/org/eclipse/smarthome/io/net/security/internal/NetworkServerTlsProviderTest.java

```diff
@@ -0,0 +1,321 @@
+        // createSSLContext will fail, because the keystore password is unknown
+        thrown.expect(IOException.class);
+        thrown.expectCause(IsInstanceOf.<Throwable> instanceOf(UnrecoverableKeyException.class));
+        provider.createSSLContext("default");
+    }
+
+    @Test
+    public void createValidSelfSignedCertificateIfNoExists()
+            throws GeneralSecurityException, IOException, NoSuchAlgorithmException, KeyManagementException,
+            InterruptedException, ExecutionException, TimeoutException {
+        SSLContext s = provider.createSSLContext("default");
+        ServerSocket serverAcceptSocket = s.getServerSocketFactory().createServerSocket();
+        serverAcceptSocket.bind(new InetSocketAddress(0));
+
+        // Create client socket
+        SSLContext sslContext = SSLContext.getInstance("TLSv1.2");
+        sslContext.init(null, new TrustManager[] { new X509ExtendedTrustManagerEx() },
+                new java.security.SecureRandom());
+        SSLSocket clientSocket = (SSLSocket) sslContext.getSocketFactory().createSocket();
+
+        final CompletableFuture<Socket> c = CompletableFuture.supplyAsync(() -> {
+            try {
+                return serverAcceptSocket.accept();
+            } catch (IOException e) {
+                throw new CompletionException(e);
+            }
+        });
+
+        // Connect
+        clientSocket.connect(serverAcceptSocket.getLocalSocketAddress());
+        SSLSocket serverSocket = (SSLSocket) c.get(200, TimeUnit.MILLISECONDS);
+
+        final CompletableFuture<Void> serverHandshake = CompletableFuture.runAsync(() -> {
+            try {
+                serverSocket.startHandshake();
+            } catch (IOException e) {
+                throw new CompletionException(e);
+            }
+        });
+
+        final CompletableFuture<Void> clientHandshake = CompletableFuture.runAsync(() -> {
+            try {
+                clientSocket.startHandshake();
+            } catch (IOException e) {
+                throw new CompletionException(e);
+            }
+        });
+
+        CompletableFuture.allOf(serverHandshake, clientHandshake).get(500, TimeUnit.MILLISECONDS);
+
+        // Check certificate
+        SSLSession session = clientSocket.getSession();
+        X509Certificate cert = (X509Certificate) session.getPeerCertificates()[0];
+        assertTrue(cert.getSubjectDN().getName().contains("O=\"Eclipse Foundation, Inc.\""));
+
+        // Send
+        clientSocket.getOutputStream().write(new byte[] { 'c' });
+        clientSocket.getOutputStream().flush();
+        serverSocket.getOutputStream().write(new byte[] { 's' });
+        serverSocket.getOutputStream().flush();
+
+        // Receive
+        clientSocket.setSoTimeout(200);
+        serverSocket.setSoTimeout(200);
+        byte buffer[] = new byte[10];
+        assertThat(clientSocket.getInputStream().read(buffer), is(1));
+        assertThat(buffer[0], is((byte) 's'));
```

Comment:
They both arrive simultaneously or to put it in another way: The OS holds a receive buffer for us

Reply:
If they both arrive at the same time, should they not be in the buffer on the first `read` from `clientSocket.getInputStream()`, i.e. in `buffer[0]` and `buffer[1]`? Because your sequential reading reads the `s` first` and then the `c`. What if they arrive the other way around?

Reply:
I set the timeout to 200ms, and send both bytes (that's async). Then I first block wait on the client side (even if the server side received its byte already) and receive and then on the server side.

The test might run a tiny bit faster if I have two receiving CompletableFutures like in the dtls tests, but creating the threads may as well slow things down.